code/controllers/subsystems/mapping.dm

```diff
@@ -4,71 +4,76 @@ SUBSYSTEM_DEF(mapping)
-		marker.load_template()
+	// Fetch and track all templates before doing anything that might need one.
+	for(var/datum/map_template/MT as anything in get_all_template_instances())
+		register_map_template(MT)
+
+	// Populate overmap.
+	if(length(global.using_map.overmap_ids))
+		for(var/overmap_id in global.using_map.overmap_ids)
+			var/overmap_type = global.using_map.overmap_ids[overmap_id] || /datum/overmap
+			new overmap_type(overmap_id)
+	// This needs to be non-null even if the overmap isn't created for this map.
+	overmap_event_handler = GET_DECL(/decl/overmap_event_handler)
 
-	preloadTemplates()
+	// Build away sites.
 	global.using_map.build_away_sites()
+
 	. = ..()
 
 /datum/controller/subsystem/mapping/Recover()
 	flags |= SS_NO_INIT
-	map_templates = SSmapping.map_templates
-	space_ruins_templates = SSmapping.space_ruins_templates
-	exoplanet_ruins_templates = SSmapping.exoplanet_ruins_templates
-	away_sites_templates = SSmapping.away_sites_templates
-
-/datum/controller/subsystem/mapping/proc/preloadTemplates(path = "maps/templates/") //see master controller setup
-	var/list/filelist = flist(path)
-	for(var/map in filelist)
-		var/datum/map_template/T = new(paths = "[path][map]", rename = "[map]")
-		map_templates[T.name] = T
-	preloadBlacklistableTemplates()
-
-/datum/controller/subsystem/mapping/proc/includeTemplate(var/datum/map_template/map_template, var/list/banned_maps)
-	if(!initial(map_template.id))
-		return
-	var/datum/map_template/MT = new map_template()
-	if(banned_maps)
-		for(var/mappath in MT.mappaths)
-			if(banned_maps.Find(mappath))
-				return
-	map_templates[MT.name] = MT
-	. = MT
+	map_templates =             SSmapping.map_templates
+	map_templates_by_category = SSmapping.map_templates_by_category
+	map_templates_by_type =     SSmapping.map_templates_by_type
 
-/datum/controller/subsystem/mapping/proc/preloadBlacklistableTemplates()
-	// Still supporting bans by filename
-	var/list/banned_exoplanet_dmms = generateMapList("config/exoplanet_ruin_blacklist.txt")
-	var/list/banned_space_dmms = generateMapList("config/space_ruin_blacklist.txt")
-	var/list/banned_away_site_dmms = generateMapList("config/away_site_blacklist.txt")
+/datum/controller/subsystem/mapping/proc/register_map_template(var/datum/map_template/map_template)
+	if(!validate_map_template(map_template) || !map_template.preload())
+		return FALSE
+	map_templates[map_template.name] = map_template
+	for(var/temple_cat in map_template.template_categories) // :3
+		LAZYINITLIST(map_templates_by_category[temple_cat])
+		LAZYSET(map_templates_by_category[temple_cat], map_template.name, map_template)
+	return TRUE
 
-	if (!banned_exoplanet_dmms || !banned_space_dmms || !banned_away_site_dmms)
-		report_progress("One or more map blacklist files are not present in the config directory!")
+/datum/controller/subsystem/mapping/proc/validate_map_template(var/datum/map_template/map_template)
+	if(!istype(map_template))
+		return FALSE
```

Comment:
May as well crash here.

Reply:
Added a stack trace - I don't think crashing during a validation proc is necessarily helpful.